[Team-16][iOS] DetailView 구현

sanghyeok-kim opened the pull request:
안녕하세요 Lena!!! 메이스와 피그백(@Piggy-Seob)입니다. 오랜만의 PR이군요 😊
지난 PR에서 초기 화면 뷰(`OrderingCollectionView`)에 대한 작업을 했다면,
이번엔 상품 상세 정보 뷰(`DetailScrollView`)를 구현하는 작업을 진행했습니다.

DetailView를 구현하는 과정에서 특이사항이 있다면
스택뷰를 정~~~~~말 많이 사용하게 되었고,
그 과정에서 '뷰 안에 스택 뷰 안에 스택 뷰 안에 뷰 ...' 같은 형태로 뷰 계층이 많이 복잡해지게 되었습니다.
그래서 뷰 계층을 파악하는데 도움을 드리고자 아래에 계층 구조를 적어두었습니다.
많이 복잡하거나 재사용이 가능한 뷰의 경우 나름대로 CustomView로 분리하고, 코드들을 함수화해서 압축시켜 봤지만
그래도 여전히 복잡하다는 점.. 미리 양해부탁드리겠습니다 😂

## Feature

- DetailView 구현

|실행 화면|
|:-:|
|![CleanShot 2022-04-27 at 18 17 53](https://user-images.githubusercontent.com/57667738/165485778-c999a4f2-2260-4a54-a686-6b2ea49e7ae2.gif)|

|||
|:-:|:-:|
|![CleanShot 2022-04-27 at 18 20 21](https://user-images.githubusercontent.com/57667738/165486271-71af3188-9d8e-4d44-bc07-3b5e36d7e425.png)|![CleanShot 2022-04-27 at 18 21 05](https://user-images.githubusercontent.com/57667738/165486392-f7227891-7d17-4c63-9d9a-e040c3c088bf.png)|

  ![스크린샷 2022-04-27 오후 5 07 06](https://user-images.githubusercontent.com/57667738/165483863-b241d05f-f7c1-424e-a3bc-0141c79e9185.png)
  - DetailScrollView
    -  ContentView
       - overViewImageScrollView (메뉴 썸네일 이미지)
       - imagePageControl
       - detailContainerStackView
         - mainInfoView (메뉴명, 메뉴 설명, 가격, 특가 뱃지) - CustomView로 분리
         - subInfoView (적립금, 배송정보, 배송비) - CustomView로 분리
         - countStackView (수량)
         - amountStackView (총 주문금액, 주문하기 버튼)
         - recipeImageStackView (레시피 이미지)

- 썸네일 이미지와 레시피 이미지를 서버로부터 받아와 출력
- Stepper를 통해 주문 수량 변경, 총 금액도 수량에 맞게 변경되도록 반영

<br>

## In Progress
- View가 URLSession을 통해 data를 직접 받아와서 그리는 방식
  -> VC가 URLSession으로 data를 받아와서 View에 넣어주는 방식으로 변경으로 진행중
  (현재 DetailViewController 부분만 우선적으로 리팩토링 된 상태입니다.)


## To Do
(`?`가 붙은 작업은 옵셔널 입니다.. 할 수 있을지 모르겠지만 도전해보겠습니다 😭)
- 현재 진행중인 리팩토링 마무리(View가 아닌 VC가 URLSession으로 데이터 받아오기)
- URLSession에서 shared가 아닌 downloadTask로 받아오기`?`
- 이미지 캐싱 구현하기
- OAuth를 적용하여 GitHub 계정으로 로그인 기능 적용`?`
- 메인반찬, 국물요리, 밑반찬 3가지 메뉴 종류를 구분해서 받아오기(현재 메인반찬으로만 모두 받아온 상태)

## 질문거리
이번 작업을 하면서 생겼던 궁금증이나 고민했던 내용들은
아래 코드에 Comment를 달아서 남기겠습니다 😀

sanghyeok-kim committed:
[Refactor][#13] deleteCase도 HTTPMethod.delegate의 description으로 변경

piggyse committed:
[fix] headerDelegate Bug

- Contents
SectionHeader 클릭시 모든 SectionHeader에 이벤트 발생 버그 제거
prepareForReuse에 CountLabel 제거 로직 추가

Commit:
[Merge pull request] PiggySeob/feature/HeaderDelegate -> Develop

[fix] headerDelegate Bug

sanghyeok-kim committed:
[Feat][#10] DetailView의 MenuStackView 구현

Contents
- DetailView에 들어갈 Menu 제목, 설명, 가격, 뱃지에 대한 뷰를 구현
- 기존 OrderingViewCell에서 사용되던 MenuStackView를 별도의 클래스 파일로 분리하여, 폰트와 스페이싱만 재조정해서 재사용하도록 구현

Todo
- 서버로부터 받아온 데이터와 연결하여 아래의 나머지 뷰 구현

sanghyeok-kim committed:
[Merge pull request]  feature/DetailViewMock -> develop

Detail View 구현 - MenuStackView 분리하여 재활용

sanghyeok-kim committed:
[Rename] UIColor+Extension 오타 수정

sanghyeok-kim committed:
[Rename][#10] menuStackView의 메소드명 불필요한 네이밍 수정

sanghyeok-kim committed:
[Feat][#10] 메뉴 부수 정보를 위한 SubInfoTitle 상수 문자열 선언 및 구현

sanghyeok-kim committed:
[Feat][#10] 메뉴 부수 정보를 출력하는 Custom StackView - MenuSubInfoContainerStackView 생성 및 구현

Content
- MenuSubInfoContainerStackView 생성
- 매개변수로 String 배열을 받아서 각 스택뷰에 description으로 넣은 뒤, 해당 스택뷰 3개를 담는 하나의 스택뷰
- Constant로 만든 SubInfoTtile의 개수보다 더 많이 입력될 경우 error

sanghyeok-kim committed:
[Feat][#10] DetailViewController 뷰 구현 (부수 메뉴 출력까지)

Content
- MockImage로 ImageScrollView 구현
- ImagePageControl 구현
- menuInfoStackView 구현
- subInfoStackView 구현

Todo
- MockImage를 서버에서 받아온 값으로 변경
- 수량 Stepper, 주문 금액 및 주문 버튼 구현
- 레시피 이미지 출력 구현

sanghyeok-kim committed:
[Merge pull request] feature/DetailViewRecipe -> develop

[#10] Detail View 구현 - MenuStackView, MenuSubInfoContainerStackView

sanghyeok-kim committed:
[Feat][#10] 총 주문 금액 및 주문 버튼 구현 (amountTitleLabel, amountLabel, orderButton)

Content
- amountTitleLabel, amountLabel을 horizental하게 갖는 amountContainerStackView 구현
- amountContainerStackView, orderButton을 vertical하게 갖는 orderContainerStackView 구현
- Stepper 동작 구현

Todo
- contentView가 detailContainerStackView의 height에 맞게 늘어나지 않는 버그 수정

piggyse committed:
Merge branch 'feature/DetailViewRecipe' of https://github.com/P-SeoB/sidedish into feature/DetailViewRecipe

piggyse committed:
[fix][#10] ContentView 스크롤 가능범위 버그 해결

Content
- 버튼이 화면을 넘어가던 문제 해결
- ContentLayoutGuide를StackView의 Bottom으로 설정 추가

sanghyeok-kim committed:
[Feat][#10] 메뉴 thumbnail 및 recipe 이미지뷰 서버로부터 받은 값으로 출력

Content
- thumbNail 이미지뷰 서버로부터 받아와서 출력
- recipe 이미지뷰 서버로부터 받아와서 출력
- recipe 이미지뷰의 비율은 유지한 채 위아래 공백만 잘라내는 Custom UIImageView Extension 구현

Todo
- Stepper를 통해 총 주문 금액 변화값 반영하기

sanghyeok-kim committed:
[Feat][#10] PriceConvertor 생성 및 구현

Content
- 'xx,xxx원' 형태의 String값을 Int로 변환하는 기능
- Int값을 'xx,xxx원' 형태의 String값으로 변환하는 기능

sanghyeok-kim committed:
[Rename][#10] Price 파일명 변경 -> PriceConvertor

sanghyeok-kim committed:
[Rename][#10] Price 파일명 변경 -> PriceConvertor

sanghyeok-kim committed:
[Feat][#10] 총 합계 금액을 계산하는 로직 구현

Content
- MenuDetail 모델 데이터를 DetailVC가 갖도록 변경
- DetailScrollView에 옵저버프로퍼티 amount 추가
- PriceConvertor 사용하여 String형태와 Int형태를 변환
- 값이 변경될 때 마다 amountLabel의 text값을 변경

sanghyeok-kim committed:
[Feat][#10] DetailScrollView 스크롤 하단에 여백의 미 추가

sanghyeok-kim committed:
[Merge pull request] feature/DetailViewRecipe -> develop

[#10] Detail View 구현

sanghyeok-kim committed:
[Simplify][#10] 불필요한 코드 제거

sanghyeok-kim committed:
[Rename] Constant 네이밍 수정 및 DetailScrollView에 대한 Constant 열거형 추가

sanghyeok-kim committed:
[Refactor] ViewController 파일 불필요한 코드 제거 및 코드 함수화

sanghyeok-kim committed:
[Refactor] SectionHeaderView 코드 함수화 및 protocol 분리

sanghyeok-kim committed:
[Merge pull request] feature/DetailViewRecipe -> develop

코드 전반 리팩토링 작업

sanghyeok-kim committed:
[Refactor] DetailScrollView 코드 함수화 리팩토링

Content
- setUpView() 내부 코드 함수화하여 정리
- backgroundColor 지정

sanghyeok-kim committed:
[Merge pull request] feature/DetailViewRecipe -> develop

[Refactor] DetailScrollView 코드 함수화 리팩토링

sanghyeok-kim committed:
[Refactor][#27] DetailScrollView에서 직접 수행하던 URLSession을 DetailVC가 수행하도록 변경(진행중)

Todo
- Decoding Error 원인 파악하여 해결

piggyse committed:
[Refactor][#27] NetworkMange 에러 해결

Contents
- Decoding할 데이터 타입이 없어서 Decoding Error 발생
- imageFecth하는 경우를 분기하여 처리.

Commit:
[Merge pull request] Piggy-Seob/migrateDetailViewNetwork - > develop

DetailViewController Decoding Error 해결

sanghyeok-kim committed:
[Rename] StackView 네이밍 수정

piggyse commented on SideDishApp/SideDishApp/Network/NetworkManager.swift:
```diff
@@ -63,17 +65,20 @@ final class NetworkManager {
+                
+                if getImageBaseURL == BaseURL.image.urlString {
+                    return completion(.success(data as? T))
                 }
```
네트워크 작업시 하나의 함수로 get,post.. 등등의 메서드를 처리하기 위해서 request라는 함수를 만들었습니다.
하지만, 네트워크에서 `별다른 Decoding할 자료가 오지 않을 경우 제네릭 T값으로 Decoding하는 쪽에서 에러`가 났습니다.
따라서, 미리 만들어 두었던 Endpoint의 속성중에 baseURL이 만약 image요청일 경우 decoding을 시도하지 않고 바로 
Success를 반환하도록 함수를 수정하였습니다.
하지만 수정후에 보니, 이대로 계속 분기해서 모든 에러를 처리한다면 결국 함수가 엄청 커질것 같다는 느낌이 들었습니다.
이런경우에, 코드의 내용이 조금 겹치더라도 Decoding을 하지않는 함수를 따로 만드는게 좋을까요 ..?
아니면, 그냥 에러Case가 생길때마다 처리를 해주는게 좋을까요??? 🥲
별개로.. 저희가 이 함수의 구현부에 `제네릭 T값에 Data타입`을 넣어주고 있는데, Data타입으로 온 서버의 data가 왜 decoding이 안되는지 아직도 의문입니다 .. ㅠㅠ😂

dev-Lena replied:
이런 경우를 위해 try 문을 사용하고 있는데요. catch로 에러가 났을 때의 예외처리도 추가하는건 어떨까요? 로그를 찍거나 기본 이미지룰 return 하도록이요.
레이어를 나누는 건 설계에 따라 다르겠지만...! 서버에서 받아온 데이터를 디코딩까지 해서 사용부에 전달하길 원하신다면 위 방법은 어떠신가요?? 😃

추가로 
``` swift
// 기존 코드
let fetchedData = try JSONDecoder().decode(T.self, from: data)
```
여기서 디코딩할 타입이 옵셔널이 아니어서 `별다른 Decoding할 자료가 오지 않을 경우 제네릭 T값으로 Decoding하는 쪽에서 에러`가 난건 아닐까 싶습니다. 👀

piggyse replied:
오..try문을 또 만들 생각은 못했네요!! 그부분도 한번 시도해 보겠습니다 감사합니다 🙃

piggyse commented on SideDishApp/SideDishApp/View/DetailScrollView.swift:
```diff
@@ -0,0 +1,338 @@
+        didSet {
+            countLabel.text = "\(orderCount)"
+        }
+    }
```
이부분은 Property Wrapper로 초기값을 설정을 해주려 했으나, Stepper의 기본 기능중에 minimumValue를 설정하는 프로퍼티가 있어서 그냥 DidSet으로 빼보았습니다!

dev-Lena replied:
이 경우는 프로퍼티 래퍼보다는 didSet이 더 적절한 것 같네요 👍🏻

piggyse commented on SideDishApp/SideDishApp/View/DetailScrollView.swift:
```diff
@@ -0,0 +1,338 @@
+        let adjustedImageHeight = baseWidth * imageRatio
+        
+        return adjustedImageHeight
+    }
```
서버에서 오는 `이미지의 크기가 들쭉날쭉해서(특히 커서) 비율을 맞추는데에 많은 고민`이 있었습니다.
특히, 비율에 맞게 fit을 하면 가로길이는 맞지만 `위아래로 비정상적으로 공백이 생기는 문제`가 생겼고 확인해 본 결과 
image의 비율에 맞춰서 줄었지만, imageView의 크기가 줄어들지 않아서 생기는 문제 같았습니다.
따라서,imageView에 image의 `width를 기준으로 비율을 만들어서 imageView의 사이즈를 resize하는 방법`을 생각해 보았습니다!

dev-Lena replied:
오 리사이즈 좋네요!! 👍🏻 이미지뷰 안에 이미지를 어떤 비율로 어떻게 표시할지 여러 옵션이 있는데 [contentMode](https://developer.apple.com/documentation/uikit/uiview/1622619-contentmode)도 한 번 확인해보시면 좋을 것 같네요 ☺️

piggyse replied:
`ContentMode`를 시도했으나 여전히 같은 문제가 발생하고 원하는 비율이 나오지않아 사용해 보았습니다 😂

sanghyeok-kim commented on SideDishApp/SideDishApp/View/DetailScrollView.swift:
```diff
@@ -0,0 +1,338 @@
+        layoutOrderButton()
+    }
+    
+    private func addPlaceholderView(count: Int) {
+        for _ in 0..<count {
+            recipeImageStackView.addArrangedSubview(UIView())
+        }
+    }
```
(이 코드를 작성하게 된 이유)

우선 이미지를 화면에 띄우는 과정은 아래와 같습니다.
1. 서버로부터 이미지 URL 주소들을 String으로 담고 있는 'String 배열'을 받아옵니다.
2. 해당 'String 배열'을 순회하며 실제 이미지 data를 받아옵니다.

1번 과정에서는 이미지의 URL 주소 String의 순서가 정렬된 상태로 오기 때문에 상관이 없지만,
2번 과정에서 해당 URL들을 통해 실제 이미지 data를 받아올 때는
이미지 크기 등의 차이에 의해 순서가 뒤바뀌어 오게 됩니다.

이를 해결하기 위해서는, 단순히 `addArrangedSubview`가 아닌,
아래 코드처럼 `insertArrangedSubview(at:)`을 사용하여 
'어느 이미지가 어느 순서에 들어가야 하는지'를 정해주어야 했습니다.
```swift
self.recipeImageStackView.insertArrangedSubview(imageView, at: imageIndex)
```

하지만 이렇게 할 경우,
n번째 보다 n+1번째 이미지가 먼저 도착하는 경우 때문에
아직 존재하지 않는 인덱스 범위에 insert를 하게되어
out-of-bound 에러가 발생하는 문제가 있었습니다.

이를 해결하기 위해
1번 과정에서 String 배열의 count를 통해 '넣어야할 이미지의 개수'를 파악하고,
해당 count만큼 빈 뷰를 placeholder 용도로 넣어두었습니다.
이렇게 할 경우 어떤 index에 먼저 insert를 하더라도
그 자리에 placeholder로 뷰가 존재하므로 out-of-bound 에러가 발생할 일이 없습니다.

---

(궁금증)

이미지를 URLSession으로 받아올 때, 이미지의 순서가 뒤죽박죽이었던 문제를 해결하기 위해 사용한 방법(꼼수)입니다.
이런 방식을 사용하는 것이 괜찮은지, 더 나은 방법이 있는지에 대한 Lena의 의견이 궁금합니다!

dev-Lena replied:
꼼수라고 하셨지만 제 생각에 이 방법도 괜찮은 방법입니다. 
순서를 유지하는 방법은 여러가지가 있는데, 아예 네트워크 요청과 데이터 수신을 순서대로 되도록 할 수도 있습니다. 
하지만 사실 이렇게까지 하면 오버 엔지니어링이 될 수 있다고 생각하는데요 혹시 궁금하시면 [OperationQueue](https://developer.apple.com/documentation/foundation/operationqueue), [Operation](https://developer.apple.com/documentation/foundation/operation#1661485) 이 문서를 참고해보시면 좋을 것 같습니다!

sanghyeok-kim commented on SideDishApp/SideDishApp/Controller/DetailViewController.swift:
```diff
@@ -6,12 +6,133 @@
+        detailScrollView.countStepper.addTarget(self, action: #selector(stepperValueChanged(_:)), for: .touchUpInside)
+    }
+    
+    private func configureMenuStackView() {
+        detailScrollView.mainInfoStackView.setTitle(by: menu.title)
+        detailScrollView.mainInfoStackView.setDescription(by: menu.description)
+        detailScrollView.mainInfoStackView.setPrice(originPrice: menu.n_price, discountedPrice: menu.s_price)
+        detailScrollView.mainInfoStackView.setBadges(by: menu.badge)
+    }
```
뷰 구조가 복잡해지면서 뷰를 계층적으로 사용하다보니 생긴 고민입니다.

DetailVC가 detailScrollView를 가지고 있고, detailScrollView가 mainInfoStackView를 가지고 있습니다.
그리고 mainInfoStackView는 자신의 뷰 요소들을 설정할 수 있는 `set~()` 함수들을 제공하고 있습니다.

이런 상황에서 DetailVC는 mainInfoStackView에 직접 접근할 수 없고,
반드시 detailScrollView를 거쳐서 접근해야 합니다.

여기서 생기는 고민은 다음과 같습니다.

1. detailVC가 mainInfoStackView의 `set~()` 함수들에 접근하기 위해서는
우선 mainInfoStackView에 접근할 수 있어야 하기 때문에
mainInfoStackView의 접근제어자를 private(set)으로 변경해야 한다는 점

2. 또한 detailVC가 `set~()` 함수들에 접근하기 위해 `.`을 2번 사용하게되면서
뭔가 객체지향의 원칙과도 멀어지게 되지 않는가.. 하는 점

여기서 제가 생각한 가장 간단한 해결 방법은,
'detailVC가 mainInfoStackView를 직접 가지면 어떨까'였습니다.
이 경우 detailVC가 mainInfoStackView에 직접 접근할 수 있으므로
`set~()`함수들에게도 `.`을 한 번만 사용해서 접근할 수 있습니다.

다만 이 경우에는 또 다른 문제가 생깁니다.
VC가 detailScrollView만 관리(addSubview + layout)하면 되던 이전의 상황과 달리,
이제는 mainInfoStackView에 대해서도 관리를 해야하므로

detailVC가 detailScrollView에 mainInfoStackView를 넣어줘야함과 동시에
detailScrollView 속에서의 mainInfoStackView의 layout까지도 직접 잡아줘야합니다.

그런데 여기서 만약 mainInfoStackView의 layout을 detailScrollView 속에 있는 또 다른 뷰들과의 관계로 잡아줘야 한다면..?
그 '또 다른 뷰들'에 대해서도 detailVC가 접근할 수 있어야 하는데..?
하는 끝없는 딜레마의 반복이 시작됩니다...ㅠㅠ

질문이 너무 길어져서 죄송합니다 ㅠㅠ
뷰를 짜면서 항상 고민거리로 작용했던 부분이라, Lena의 의견을 듣고 싶습니다.

---

A뷰 안에 들어가는 B뷰가 있을 때, VC가 B뷰의 함수에 접근하기 위해서 어떻게 해야 할까?
1. A뷰가 B뷰를 관리하고, VC는 A뷰를 거쳐서 B뷰의 함수에 접근한다.(`A뷰.B뷰.B함수`)
2. A뷰와 B뷰를 모두 VC가 관리하고, VC는 B뷰의 함수에 직접 접근한다.(`B뷰.B함수`)
  단, 이 경우 VC가 A뷰 안에 B뷰를 addSubview해줘야하며, A뷰 속에서의 B뷰의 layout도 VC가 잡아줘야 한다.
3. A가 B뷰를 관리하는 방식은 유지하면서, A뷰가 'B뷰의 함수에 접근하는 함수'를 제공한다.
  (`A뷰.A함수` -> A함수는 B함수를 호출)
  ```swift
    //3번 방식
    A뷰.setB뷰의title(text: String) {
      B뷰.setTitle(text: text)
    }
  ```

dev-Lena replied:
이 질문이 View 계층이 깊고, 상위 계층에서 하위 계층으로 뷰에 들어갈 데이터를 제공해 View를 그려야하는 경우, 어떻게 하는게 좋을지 저의 의견을 묻는거라고 이해했는데 맞을까요? 👀

지금 코드를 보니
`OrderingVC`에서 서버에서 가져온 데이터를 `DetailViewController`의 init 시점에 주입을 받네요.
저의 의견을 드리자면 주입 받은 데이터를 폭포수처럼 하위 계층의 뷰에게 전달만 하고 그 데이터를 기반으로 UI를 그리는건 각 뷰에서 하도록 하는건 어떤가요?
그러면 영혼 깊숙한 곳에서 끌어오는 것 같은 메서드 콜도 없어도 되고 VC가 모든 뷰를 직접 관리하거나 그리지 않아도 됩니다. 그리고 데이터를 넘겼을 때 기왕이면 데이터 변경 됐을 때 이 타이밍을 캐치해서 뷰가 업데이트 되면 좋겠네요. ☺️

sanghyeok-kim committed:
[Refactor][#27] DetailScrollView에서 직접 수행하던 URLSession을 DetailVC가 수행하도록 변경(완료)

piggyse committed:
[Refactor][#22] HeaderConstant 제거

piggyse committed:
[Refactor][#22] Category Enum에 HeaderValue 추가

piggyse committed:
[Refactor][#22] Section별 네트워크 Request

Content
- Main메뉴 뿐만 아니라 Section별로 메뉴 가져오기
- CollectionView DataSource index(Int) -> indexPath로 변경
- Configure 함수 매개변수 index -> menu로 변경
- dish -> menu 매개변수명 통일

Commit:
[Merge pull request] feature/URLSession -> develop

Section별 네트워크 Request작업. #22

piggyse committed:
[rename] dish -> menu

piggyse committed:
[Refactor] Cell에서 Fecth이미지 하던 로직 DataSource로 이전

Commit:
[Merge pull request] feature/MenuImageFecthRefactory -> develop

[Refactor] Cell에서 Fecth이미지 하던 로직 DataSource로 이전

dev-Lena reviewed (approved):
> 안녕하세요 Lena!!! 메이스와 피그백(@Piggy-Seob)입니다. 오랜만의 PR이군요 😊
지난 PR에서 초기 화면 뷰(OrderingCollectionView)에 대한 작업을 했다면,
이번엔 상품 상세 정보 뷰(DetailScrollView)를 구현하는 작업을 진행했습니다.
DetailView를 구현하는 과정에서 특이사항이 있다면
스택뷰를 정~~~~~말 많이 사용하게 되었고,
그 과정에서 '뷰 안에 스택 뷰 안에 스택 뷰 안에 뷰 ...' 같은 형태로 뷰 계층이 많이 복잡해지게 되었습니다.

다시 반갑습니다 메이스와 피그백 ☺️~~~~
가슴이 웅장해지는 구조군요 🙈 고생하셨습니다. 
각 코드에 이유가 있어서 좋네요 ☺️👍🏻👍🏻 
질문을 구체적으로 주셔서 질문을 파악하기가 좋았던 것 같아요! 
각 부분의 질문에 대한 답변이나 의견 드렸습니다.

추신. (혹시나) 제가 확인하고 리뷰한 코드와 변경된 코드가 다를 수 있어 혹오옥시 저를 리뷰어로 지정한 후에 변경사항을 커밋했다면 어떤 부분을 수정했는지 코멘트로 알려주시면 감사합니다. ☺️ 

❗️ 와우 제가 마감시간을 착각했네요. 리뷰 늦어서 죄송합니다. 😓 그래서 여기에 커밋을 붙이신 거였군요 ㅠㅠ
